ARCHITECTURE.md

```diff
@@ -0,0 +1,162 @@
+
+To counter this, we cache the reference to Android UI thread's classloader, and use it to lookup classes (instead of `FindClass`).
+
+Flutter embedding also doesn't allow us to return from FFI when a JNI exception is pending. To counter this, we usually return a `JniResult` struct. It will contain any exceptions occurred during the call. It's checked using extension getters in Dart layer. If any exception has occurred, using some helper methods, we convert both the Java stack trace and message of the exception into dart strings, and rethrow in Dart.
+
+### On standalone
+With `Jni.spawn` it's usually necessary to specify classpaths. ClassPath is nothing but another JVM option. It has some additional code handling it in [jni.dart](jni/lib/src/jni.dart) so that classpaths can be passed as a dart list of paths.
+
+`jni:setup` scans the packageConfig and finds JNI plugins by looking for the first line of `CMakeLists.txt`. It should be identical to that of [jni/src/CMakeLists.txt](jni/src/CMakeLists.txt). `jnigen` ensures this while generating the CMake configuration for generated bindings.
+
+### Bindings to `jni.h`
+`package:jni` also contains some automatically generated bindings to `jni.h`. This header file is taken from Android NDK and annotated with some parameter names. The bindings are generated using a patched and vendored version of `ffigen`. The main purpose of the patch is to generate extension methods for function pointer fields. We use this idiom in both `GlobalJniEnv` and `JniAccessors` types. It also hides some pointless methods, for instance the ones taking `va_list` parameters.
+
+### CMake configuration
+CMake configuration for `package:jni` is pretty simple. It uses CMakes builtin detection of JNI libraries. Only thing worth noting is the `DELAYLOAD` flag when linking `jvm.dll` on Windows.
+
+## ApiSummarizer
+ApiSummarizer is an important component of `jnigen`. It's a pretty standard maven Java project. We use `assembly` plugin for building a runnable JAR. The confusing details of building it are in the dart code! It generates an API tree in JSON format.
+
+ApiSummarizer contains 2 "back-ends". Doclet is OpenJDK's API which enables us to plug into JavaDoc tool, and create description of the public API provided by source. We use ToolProvider API so that we can invoke JavaDoc in-process rather than as external command.
+
+Most of ApiSummarizer is pretty boring methods, which scan the JavaDoc's representation of the API, and serialize it into a tree of plain data types. This can be easily written to stdout using jackson-databind. On the dart side, we have [elements.dart](jnigen/lib/src/elements/elements.dart) which defines the same structures. JSON deserialization code is generated using `json_serializable` code generation package.
```

Comment:
Most of ApiSummarizer is -> are

Reply:
This is hard grammar decision. I prefer `A large part is`.